### PR TITLE
[SUPERSEDED] Review/repartee [ci: last-only]

### PR DESCRIPTION
**Pull request by som-snytt**
Unsquashed version of the reporting change, with some tweaks to point out.

Basically alike, but maybe a shift in philosophy toward supporting the non-filtering client. Filtering is such a hassle! How about you do the filtering and just hand me events to work on.

I have a concern that I'll want to write a reporter that sees everything and I can't because the event stream is pre-filtered. I haven't had time to do that exercise yet.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala [hunk 1]
```diff
@@ -235,22 +234,18 @@ abstract class BCodeHelpers extends BCodeIdiomatic {
...
+   * global.reporter to suppress reportage.
    */
   def completeSilentlyAndCheckErroneous(sym: Symbol): Boolean =
     if (sym.hasCompleteInfo) false
     else {
-      withoutReporting(sym.info)
-      sym.isErroneous
+      reporter.silently {
+        sym.info
+        sym.isErroneous
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala:245]: @lrytz I haven't thought about this, but the REPL has so many ways to go silent.

File: src/compiler/scala/tools/nsc/reporters/StoreReporter.scala [hunk 2]
```diff
@@ -21,6 +21,8 @@ import scala.reflect.internal.util.Position
 class StoreReporter(val settings: Settings) extends FilteringReporter {
   import StoreReporter._
 
+  def this() = this(new Settings)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/reporters/StoreReporter.scala:24]: So here's the thing, usually I just don't care. With a plain Reporter, I can let the filter-forward thing just do it, it's essentially the compiler's responsibility, I just get the info it chooses to emit (as per normal).

File: test/junit/scala/tools/nsc/reporters/PositionFilterTest.scala [hunk 3]
```diff
@@ -16,12 +16,11 @@ class PositionFilterTest {
   val batchFile = new BatchSourceFile(source, "For testing".toList)
   val pos = new OffsetPosition(batchFile, 4)
 
-  val store = new StoreReporter
+  val store = new StoreReporter(new Settings)
```
Comment (som-snytt) [inline test/junit/scala/tools/nsc/reporters/PositionFilterTest.scala:19]: Obviously on second thought I'd have stayed with it.

File: src/compiler/scala/tools/nsc/PipelineMain.scala [hunk 4]
```diff
@@ -564,10 +564,9 @@ class PipelineMainClass(argFiles: Seq[Path], pipelineSettings: PipelineMain.Pipe
                 Position.range(sourceFile, diagnostic.getStartPosition.toInt, diagnostic.getPosition.toInt, diagnostic.getEndPosition.toInt)
               }
               diagnostic.getKind match {
-                case Kind.ERROR => reporter.error(position, msg)
+                case Kind.ERROR                            => reporter.error(position, msg)
                 case Kind.WARNING | Kind.MANDATORY_WARNING => reporter.warning(position, msg)
-                case Kind.NOTE => reporter.info(position, msg, true)
-                case Kind.OTHER => reporter.echo(position, msg)
+                case Kind.NOTE | Kind.OTHER                => reporter.echo(position, msg)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/PipelineMain.scala:569]: Nice formatting!

File: src/partest/scala/tools/partest/package.scala [hunk 5]
```diff
@@ -24,6 +24,8 @@ package object partest {
   type PathResolver = scala.tools.util.PathResolver
   type StringWriter = java.io.StringWriter
 
+  type Reporter     = scala.tools.nsc.reporters.FilteringReporter
```
Comment (som-snytt) [inline src/partest/scala/tools/partest/package.scala]: Interestingly, a bad idea.

File: src/compiler/scala/tools/nsc/reporters/Reporter.scala [hunk 6]
```diff
@@ -13,22 +13,22 @@
...
 import scala.annotation.unused
+import scala.collection.mutable
+import scala.reflect.internal, util.Position, util.ScalaClassLoader
 
 /** This class exists for sbt compatibility. Global.reporter holds a FilteringReporter.
  *  The only Reporter that is *not* a FilteringReporter is the one created by sbt.
  *  The Global.reporter_= setter wraps that in a delegating [[MakeFilteringForwardingReporter]].
  */
-abstract class Reporter extends InternalReporter {
+abstract class Reporter extends internal.Reporter {
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/reporters/Reporter.scala:24]: I wound up preferring this expression, like `mutable.Map`. It avoids indirection.

File: src/reflect/scala/reflect/internal/Reporting.scala [hunk 7]
```diff
@@ -90,13 +90,21 @@ abstract class Reporter {
...
 
+  // TODO: rename to `doReport`, remove the `force` parameter.
+  // Note: `force` is ignored. It used to mean: if `!force`, the reporter may skip INFO messages.
+  // If `force`, INFO messages were always printed. Now, INFO messages are always printed.
   protected def info0(pos: Position, msg: String, severity: Severity, force: Boolean): Unit
 
+  // 0: count and display
+  // 1: count only, don't display
+  // 2: don't count, don't display
+  def filter(pos: Position, msg: String, severity: Severity): Int = 0
```
Comment (som-snytt) [inline src/reflect/scala/reflect/internal/Reporting.scala:106]: Still on the fence about how to document, whether to use an enum. It's not an internal Rex Kerr counter thing, it's API. And there are three states, as when one speaks of the "tri-state area", everyone knows you mean New York and specifically Manhattan, no one ever goes out to 1 and 2 except to visit relatives.

File: src/compiler/scala/tools/nsc/reporters/Reporter.scala (hunk 6 above)
Comment (lrytz) [reply src/compiler/scala/tools/nsc/reporters/Reporter.scala:24]: It also works better in IntelliJ https://youtrack.jetbrains.com/issue/SCL-16096

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/reporters/Reporter.scala:24]: Ha. Maybe I'll try a contribution over there. After all, that's how Jason got his start.

**Comment by som-snytt** [comment]
I wonder if silent mode made it in.

**Comment by lrytz** [comment]
> silent mode made it in

no sorry :-) I'm trying to keep the Reporters interface light because everything added there needs to be deatl with in the ForwardingReporter, and can potentially break other custom subclasses...

**Pull request closed without merging**